include/device_queue.h

```diff
@@ -59,9 +64,166 @@ namespace detail {
 		std::unique_ptr<cl::sycl::queue> sycl_queue;
 		bool device_profiling_enabled = false;
 
-		cl::sycl::device pick_device(const config& cfg, cl::sycl::device* user_device) const;
 		void handle_async_exceptions(cl::sycl::exception_list el) const;
 	};
 
+	// Try to find a platform that can provide a unique device for each node using a device selector.
+	template <typename DeviceT, typename PlatformT, typename SelectorT>
+	bool try_find_device_per_node(std::string& how_selected, DeviceT& device, const std::vector<PlatformT>& platforms, const host_config& host_cfg,
+	    SelectorT selector, cl::sycl::info::device_type type) {
+		std::vector<std::tuple<DeviceT, size_t>> devices_with_platform_idx;
+		for(size_t i = 0; i < platforms.size(); ++i) {
+			auto&& platform = platforms[i];
+			if(platform.get_devices().size() >= host_cfg.node_count) {
+				for(auto device : platform.get_devices()) {
+					if(selector(device) == -1) { continue; }
+					devices_with_platform_idx.emplace_back(device, i);
+				}
+			}
+		}
+
+		std::sort(devices_with_platform_idx.begin(), devices_with_platform_idx.end(),
+		    [selector](const auto& a, const auto& b) { return selector(std::get<0>(a)) > selector(std::get<0>(b)); });
+		bool same_platform = true;
+		if(devices_with_platform_idx.size() >= host_cfg.node_count) {
+			auto [device_from_platform, idx] = devices_with_platform_idx[0];
+			auto platform_name = device_from_platform.get_platform().template get_info<cl::sycl::info::platform::name>();
```

Comment:
Could be made `const`. Comparing based on name seems unnecessarily brittle; `sycl::platform` implements `operator==`.

Reply:
Agreed and modified.